File: Semestrul 3/MAP/lab 2 MAP/decorator/AbstractTaskRunner.java
package sem1_2.decorator;

import sem1_2.model.Task;

//abstract decorator
public abstract class AbstractTaskRunner implements TaskRunner {
    private final TaskRunner taskRunner;

    public AbstractTaskRunner(TaskRunner taskRunner) {
        this.taskRunner = taskRunner;
    }

    @Override
    public void executeOneTask() {
        taskRunner.executeOneTask();
    }

    @Override
    public void executeAll() {
        while (hasTask()) {
            executeOneTask();
        }
    }

    @Override
    public void addTask(Task t) {
        taskRunner.addTask(t);
    }

    @Override
    public boolean hasTask() {
        return taskRunner.hasTask();
    }
}

File: Semestrul 3/MAP/lab 10 MAP/decorator/AbstractTaskRunner.cs
namespace seminar10.decorator;
using model;

public class AbstractTaskRunner(ITaskRunner taskRunner) : ITaskRunner
{
    public virtual void ExecuteOneTask()
    {
        taskRunner.ExecuteOneTask();
    }

    public void ExecuteAll()
    {
        while (HasTask())
        {
            taskRunner.ExecuteOneTask();
        }
    }

    public void AddTask(Task t)
    {
        taskRunner.AddTask(t);
    }

    public bool HasTask()
    {
        return taskRunner.HasTask();
    }
}